state-chain/pallets/cf-environment/src/lib.rs

```diff
@@ -393,6 +407,39 @@ pub mod pallet {
 
 			Ok(dispatch_result)
 		}
+
+		/// Updates the Durable Nonce for the nonce account upon witnessing that the corresponding
+		/// solana tx was successfully included in a block
+		///
+		/// ## Events
+		///
+		/// - [DurableNonceSetForAccount](Event::DurableNonceSetForAccount)
+		///
+		/// ## Errors
+		///
+		/// - [BadOrigin](frame_support::error::BadOrigin)
+		/// - [NonceAccountNotBeingUsedOrDoesntExist](Error::NonceAccountNotBeingUsedOrDoesntExist)
+		#[pallet::call_index(6)]
+		// Todo
+		#[pallet::weight(Weight::zero())]
+		pub fn update_sol_nonce(
+			origin: OriginFor<T>,
+			nonce_account: SolAddress,
+			durable_nonce: SolHash,
+		) -> DispatchResultWithPostInfo {
+			T::EnsureWitnessed::ensure_origin(origin)?;
+
+			if let Some(_nonce) = SolanaUnAvailableNonceAccounts::<T>::take(nonce_account) {
+				SolanaAvailableNonceAccounts::<T>::append((nonce_account, durable_nonce));
+				Self::deposit_event(Event::<T>::DurableNonceSetForAccount {
+					nonce_account,
+					durable_nonce,
+				});
+				Ok(().into())
+			} else {
+				Err(Error::<T>::NonceAccountNotBeingUsedOrDoesntExist.into())
```

Comment:
I'm not sure we need this error case. 

We witness the nonce accounts, so if the engines witness a new nonce account (imagine we add a new one via governance - not sure if this is possible/likely?) then we have to 'believe' them, right?

Reply:
If we were to add a nonce via governance it's unclear to me if we'd do it that way or we'd add another governance function that adds a new nonce, not just have the engine suddenly witness a new account that the SC is not aware of. But we haven't thought about that yet, so it might make sense to not throw an error as there might be a use for it.

Reply:
We have a predifined list of nonce accounts and so so if we witness a nonce for a nonce account that does is not registered with the state chain, we should ignore it. If we add a new nonce account, it should be via governance. We should have the state chain control what the nonce account list is and the cfes should witness those. We shouldnt have the cfes deciding that we now need a new nonce account and so lets start witnessing that. We should beleive the cfes for what the nonce is, not what nonce accounts to witness.

Reply:
Ok. Which account has the authority to add new durable nonce accounts? The aggkey I guess?

Reply:
Ok then I agree it makes sense to restrict this on the SC 👌